### PR TITLE
mystical agriculture plants

### PR DESCRIPTION
**Pull request by Warbringer12**
did the mystical agriculture plants I know that the last texture in the these are broken because of the way how mystical handles his plant textures

### CODE REVIEW

File: mod_mysticalagriculture/plants/air_plant.json [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+  "enabled": true,
+  "mods": [
+    "mysticalagriculture",
+    "agricraft",
+    "minecraft"
+  ],
+  "id": "mysticalagriculture:air_plant",
+  "plant_lang_key": "block.mysticalagriculture.air_crop",
+  "seed_lang_key": "item.mysticalagriculture.air_seeds",
+  "desc_lang_key": "agricraft.plant.mysticalagriculture_air.desc",
```
Comment (nbrichau) [inline mod_mysticalagriculture/plants/air_plant.json:14]: As you are using lang keys from Agricraft, could you create a PR in Agricraft adding the values for those in the lang file please.

File: mod_mysticalagriculture/soil/prudentium_farmland_soil.json [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  "version": "1.16.4",
+  "json_documentation": "https://agridocs.readthedocs.io/en/master/agri_soil/",
+  "enabled": true,
+  "mods": [
+    "mysticalagriculture",
+    "agricraft",
+    "minecraft"
+  ],
+  "id": "prudentium_farmland_soil",
+  "lang_key": "agricraft.soil.prudentium.name",
```
Comment (nbrichau) [inline mod_mysticalagriculture/soil/prudentium_farmland_soil.json:12]: I think you can use mystical agriculture lang keys for the soils.

**Review by nbrichau — CHANGES_REQUESTED** [review]
Thanks for your PR.
Some little changes are needed though.
- You can increase the tiers for all mystical agriculture and agradditions crops, at least tier 4. You can take example with agricraft resource crops.
- The mutations for the crops are missing.

But I don't think it will be merged until the last texture problem is fixed. A new render type in Agricraft need to be created for this. Moreover, the seeds are uncolored and have to be fixed somehow.

**Comment by InfinityRaider** [comment]
Indeed, thanks!
To complete @nbrichau:
1) have you tried using the models instead of an agricraft render type, does that fix the last growth stage?
2) I don't think mutations are needed for this mod since MA has its own progression system which would be broken by adding mutations.

File: mod_mysticalagradditions/plants/dragon_egg_plant.json [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    12,
+    14,
+    16
+  ],
+  "harvestStage": 4,
+  "growth_chance": 0.65,
+  "growth_bonus": 0.025,
+  "bonemeal": true,
+  "tier": 2,
+  "cloneable": true,
```
Comment (InfinityRaider) [inline mod_mysticalagradditions/plants/dragon_egg_plant.json:42]: I don't think any of these crops should be cloneable

File: mod_mysticalagradditions/plants/dragon_egg_plant.json [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    8,
+    10,
+    12,
+    14,
+    16
+  ],
+  "harvestStage": 4,
+  "growth_chance": 0.65,
+  "growth_bonus": 0.025,
+  "bonemeal": true,
```
Comment (InfinityRaider) [inline mod_mysticalagradditions/plants/dragon_egg_plant.json:40]: Are mystical agriculture crops bonemeal-able? Just asking as a check, as I'm not sure.
If they are you can ignore this comment, if they aren't, neither should they be in here

File: mod_mysticalagradditions/plants/dragon_egg_plant.json [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+      "type": "fluid",
+      "object": "minecraft:empty",
+      "useTag": false,
+      "data": "",
+      "ignoredData": []
+    }
+  },
+  "callbacks": [],
+  "texture": {
+    "render_type": "hash",
```
Comment (InfinityRaider) [inline mod_mysticalagradditions/plants/dragon_egg_plant.json:101]: Wouldn't "plus" or "cross" be more suitable?

File: mod_mysticalagradditions/plants/dragon_egg_plant.json [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    "plant_models": [],
```
Comment (InfinityRaider) [inline mod_mysticalagradditions/plants/dragon_egg_plant.json:102]: Have you tried using this?
If so I suppose it didn't work?

File: mod_mysticalagradditions/plants/dragon_egg_plant.json (hunk 5 above)
Comment (InfinityRaider) [reply mod_mysticalagradditions/plants/dragon_egg_plant.json:101]: In any case, I will be adding a mystical agriculture render type which should fix the final stage.

File: mod_mysticalagriculture/soil/prudentium_farmland_soil.json (hunk 2 above)
Comment (Warbringer12) [reply mod_mysticalagriculture/soil/prudentium_farmland_soil.json:12]: It auto gets the lang key from mystical cause it seems to keep name just fine

File: mod_mysticalagradditions/plants/dragon_egg_plant.json (hunk 6 above)
Comment (Warbringer12) [reply mod_mysticalagradditions/plants/dragon_egg_plant.json:103]: fixed all of the id names and changed tier level to 4. I know that mystical agriculture can call on plant models but it doesn't work and if you try to add in as the last texture as "mysticalagriculture:block/resource_crop_7", "mysticalagriculture:block/flower_ingot" or some other flower type like flower_rock, flower_face, flower_dust, or any other flower it doesn't render it

File: mod_mysticalagradditions/plants/dragon_egg_plant.json [hunk 7]
```diff
@@ -0,0 +1,133 @@
+{
+  "path": "mod_mysticalagradditions/plants/dragon_egg_plant.json",
+  "version": "1.16.4",
+  "json_documentation": "https://agridocs.readthedocs.io/en/master/agri_plant/",
+  "enabled": true,
+  "mods": [
```
[the remark this answers is not in the document]
Comment (Warbringer12) [reply mod_mysticalagradditions/plants/dragon_egg_plant.json:6]: add in the mod id to list of mods. the id inside in the "id": area doesn't need to be changed do to the way he calls on his plants

**Comment by InfinityRaider** [comment]
Okay, we now have added some code to support Mystical Agriculture dynamic coloring.
For this , some small modifications to the jsons are necessary though:
 - The render type should be set to "mysticalagriculture" for all plants
 - The last texture in the textures array should contain two textures: one for the stem, and another for the flowers: 
 `[
        "mysticalagriculture:block/mystical_resource_crop_7",
        "mysticalagriculture:block/flower_ingot"
      ]`

**Comment by Warbringer12** [comment]
changed render type to mysticalagriculture
changed last textures array to [ "mysticalagriculture:block/mystical_resource_crop_7", "mysticalagriculture:block/flower_ingot" ] or some other variant depending flower type

**Comment by InfinityRaider** [comment]
Perfect thanks, if @Ketheroth  doesn't have any more remarks this can be merged.

**Review by Ketheroth — CHANGES_REQUESTED** [review]
Some plants should use "flower_ingot" but doesn't use this texture. I noted some of the files but there are more. 
Can you use spaces instead of tabs for the flower texture, just to be consistent with the rest of the file.

The plants for mobs use the texture "mystical_resource_mob_\<stage\>" but should be "mystical_mob_crop\<stage\>"

Unfortunately, the seed lang keys you wrote don't exist in mystical agriculture lang file.
Btw, the `%s` mystical agriculture uses in its lang file isn't a vanilla thing, it's something they added themselves. "tooltip.mysticalagriculture.crop_id" may not be the best idea.

The fertilizers json has been changed in the meantime, the jsons will have to be adapted :
1. remove the fertilizer line in the plant jsons
2. add the plant ids to the positively affected blacklist in the bonemeal fertilizer
3. add a new fertilizer json for the mystical agriculture one.

For more information about the fertilizer json : https://agridocs.readthedocs.io/en/master/agri_fertilizer/

If I may add something else, please read the debug.log after launching the game, it will tell you if, and where, things are wrong.

**Comment by ArthixFR** [comment]
Hi, I tested these JSONs on a self-build Agricraft with MysticalAgriculture support and there are some problems so I made some changes :

- `plant_lang_key` => Incorrect and replaced with `crop.mysticalagriculture.air` (For the air seed of course)
- `seed_lang_key` => `item.mysticalagriculture.air_seeds` doesn't exist so I changed with one added in lang file (I'll commit this)
- `object` in `seed_items` => The mod ID doesn't have an underscore
- `stages` => After stage 12 the texture is flipped, and stage 5 crashes the game, so I replaced 2, 4, 6, 8, 10, 12, 14, 16 with 2, 3, 4, 5, 6, 8, 10, 12.
- `harvestStage` => I changed it to 3, can revert back to 4 but I think it's a little bit OP lol.
- `conditions` => I added crux block for MysticalAgradditions seeds

I don't know if I can make these changes here since already done all these changes on my end. Maybe creating an other pull request ?

**Comment by Ketheroth** [comment]
The texture flip was a bug that I fixed, but I have not pushed it yet in the main repo. You can rollback the changes for the stages height.
The rest seems good to me.
I think you can create a PR on Warbringer12 repo, so your changes will be here when he merge the pr. Or you can create a new PR in this repo.

**Comment by ArthixFR** [comment]
Okay ! I think I will create a PR on @Warbringer12 repo, if he's ok, It will be easier for you I think.

**Comment by Warbringer12** [comment]
That works for me so go ahead

File: mod_mysticalagriculture/plants/air_plant.json [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    2,
+    4,
+    6,
+    8,
+    10,
+    12,
+    14,
+    16
+  ],
+
```
Comment (Ketheroth) [inline mod_mysticalagriculture/plants/air_plant.json:37]: Seems like there is a \n that shouldn't be here (same for every files)

Comment (ArthixFR) [reply mod_mysticalagriculture/plants/air_plant.json:37]: Oh, do I need to change it since it's merged ?

Comment (Ketheroth) [reply mod_mysticalagriculture/plants/air_plant.json:37]: Do as you want

**Pull request merged**